### DISCUSSION TITLE
Android Studio 2023+ update and more improvements

### DISCUSSION THREAD

**Discussion by emandtf** [osmdroid/osmdroid#1963] [Show and tell]
Greetings to all,
I've forked original Master branch few months ago and choosed to add few features/improvements to the code.

What's new:
- **whole project updated to JDK17, AGP 7.4+ and Android Studio 2023+**
  MinSdkVersion is raised from 8 to 21 due to the new Lifecycle behaviour _(read more ahead)_ which is not present in oldest SDK.
  SDK 21 is Android 5.0 released on 2014 and I think oldest things could be wiped-out from code to be able to use newest techologies.
- **libraries/dependancies updated**
- **a lot of Memory footprint improvements**:
  there were a lot of "_new()_" call for Objects creation without the possibility to "_reuse_" them, so I implemented a lot o "reusable" objects. This improves also performances (because there is an overhead for each new Object instantiation in Java) especially because GarbageCollector will not run so often.
- **stop manage MapView/Overlay lifecycle events**:
  MapView is now following its container Lifecycle (as Fragments do relatively to their Activities). Calling its "_onPause()_" and "_onResume()_" is not more required. If Activity/Fragment is paused/stopped/started/resume all normal calls are ensured internally. The same is done for "_onDestroy()_" that will call "_detach()_" and "_destroy()_" as previously done.
- **stop load/download Tile when the Projection view bounds goes away**:
  previously there was a procedure that enqueues Tiles that need to be loaded/downloaded and when a Max-Queue-Length (40 if I well remember) was reached the old queued Tiles were removed from the queue. I changed this behaviour by removing Tiles from queue only when them goes outside from current View boundaries/zoom.
- **_(in MapEventsAsync new Class)_ added Callback events for offload working on a different Thread**:
  some Developer could want use some Events to load resources/layers/overlays without the need to manage Thread. At this time I've implemented a Callback system for _ON_SCROLL_ and _ON_ZOOM_ events and exposed two methods for both of them:

  > MapEventsAsync.addCallbackWorkingWithResult(CALLBACKWORKINGWITHRESULTTYPE_ONSCROLL, new CallbackWorkingWithResult<**Overlay**>() {
            [executed on a WorkerThread]
            public **Overlay** _onWorking_(MapView mapView, Object... args) { return new Overlay(....); }
            [executed on UiThread | MainThread]
            public void _onDone_(MapView mapView, **Overlay** workingResult) { ... }
    }

  More "_CALLBACKWORKINGWITHRESULTTYPE_..._" events can be associatedTo/created by _osmdroid_ Developers if needed in future versions.
- **_(in MapEventsAsync new Class)_ Async events**:
  using the new "_MapEventsAsync_" Class it's possibile to get some new MapView Events.
  Most important are:
  - _onTileLoading_
  - _onTileLoadSuccess_
  - _onTileLoadFail_
  - _onTileLoadCancelled_
  - _onViewBoundingBoxChanged_ (it's based on changes of _Tile_id_ Rect, not pixels nor lat/lon coordinates changes)
  These events are ASYNC so are raised in any moment AFTER their referral-action is already done (pay attention that those callbacks are executed on Main/UI Thread).

Included projects are well updated according to newer version allow them to be built and used without issues.

The already included "**_osmdroid-simple-map_**" example project was modified by adding an _Infinite Horizontal Progressbar_ (using newer "_MapEventsAsync_" Class) while Tiles are loading:
![20231221_progressbar_example](https://github.com/osmdroid/osmdroid/assets/13537385/45bfb00b-c923-4e4e-be0e-945dc4301051)

It could be even possibile getting how much times it tooks to load that Tile and which TileProvider is used for that specific Tile.
These are few logs while navigate through map:

> Tile 74,51,7 [SQLCACHE] loaded successfully in 7ms
> Tile 75,49,7 [SQLCACHE] loaded successfully in 6ms
> Tile 75,52,7 [DOWNLOADER] loaded successfully in 1894ms
> Tile loading cancelled 74,57,7 [DOWNLOADER] after 1902ms!!

All new visible features (not performance/memory improvements) can be tested running included "**_osmdroid-simple-map_**" example project and are available using newer _MapEventsAsync_ Class ONLY, so you need to instantiate it as shown in that example.

These things should work fine even using "_osmbonuspack_" module/library, but I've not tested all its features.

**2024-07-12: I just create a PR on master providing all those things** _(I know there are a lot of changes but it wasn't possibile without all of them)_

Best regards,
-emandtf-

**Comment by MKergall** [comment]
Interesting!

I think that everybody is allowed to create a PR in osmdroid repo.

**Reply by emandtf** [reply]
PR can be created but I cannot create a new Branch, so the only branch I could choose is Master...and I suppose those changes need to be tested a lot on a different Branch without touching Master in the near future.

**Comment by spyhunter99** [comment]
exciting! just open a PR targeting osmdroid/master
